### PR TITLE
[PM-1339] Allow Rotating Device Keys

### PR DESCRIPTION
**Pull request by justindbaur**
## Type of change

```
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

Add client side components for rotating the trusted device keys of the current device.

Related server PR: bitwarden/server#3096

## Code changes

- **change-password.component.ts:** Refactor to deconstruct the tuple to its named elements and add a call to the `DeviceTrustCryptoService` to facilitate the rotation of the current device keys so that it can stay trusted.
- **update-devices-trust.request.ts:** Added API models for the request body that is sent to the server to rotate a users devices trusted device keys.
- **protected-device.response.ts:** Added a model that returns device information including it's more sensitive parts. 
- **device-trust-crypto.service.implementation.ts:** Added logic for getting and rotating the keys of the current users device so that once logged out they can re-auth with SSO and still keep trust.
- **device-trust-crypto.service.spec.ts:** Added test asserting expected behavior.

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

### CODE REVIEW

**Review by djsmith85 — CHANGES_REQUESTED** [review]
@justindbaur 
Could you please move the following files into auth-owned folders:
`libs/common/src/abstractions/devices/devices-api.service.abstraction.ts`
`libs/common/src/abstractions/devices/requests/update-devices-trust.request.ts`
`libs/common/src/abstractions/devices/responses/protected-device.response.ts`
`libs/common/src/services/devices/devices-api.service.implementation.ts`
and likely also 
`apps/web/src/app/settings/change-password.component.ts`, but please verify with @jlf0dev on the last one

File: apps/web/src/app/settings/change-password.component.ts [hunk 1]
```diff
@@ -220,23 +222,23 @@ export class ChangePasswordComponent extends BaseChangePasswordComponent {
   }
 
   private async updateKey(masterKey: MasterKey, masterPasswordHash: string) {
-    const userKey = await this.cryptoService.makeUserKey(masterKey);
+    const [newUserKey, newProtectedUserKey] = await this.cryptoService.makeUserKey(masterKey);
```
Comment (JaredSnider-Bitwarden) [inline apps/web/src/app/settings/change-password.component.ts:225]: Nit pick (non blocking): `newProtectedUserKey` is generic, but we can be more clear by being a tiny bit more verbose with something like `masterKeyEncUserkey`

File: apps/web/src/app/settings/change-password.component.ts [hunk 2]
```diff
@@ -220,23 +222,23 @@ export class ChangePasswordComponent extends BaseChangePasswordComponent {
...
     const privateKey = await this.cryptoService.getPrivateKey();
```
Comment (JaredSnider-Bitwarden) [inline apps/web/src/app/settings/change-password.component.ts:226]: Nit pick (non blocking): `userPrivateKey` ?

File: apps/web/src/app/settings/change-password.component.ts (hunk 1 above)
Comment (JaredSnider-Bitwarden) [reply apps/web/src/app/settings/change-password.component.ts:225]: Love the clarity you've added here with the array destructuring and `new` prefix!

File: libs/common/src/abstractions/devices/responses/protected-device.response.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    this.encryptedPublicKey = this.getResponseProperty("encryptedPublicKey");
+  }
+
+  id: string;
+  name: string;
+  type: DeviceType;
+  identifier: string;
+  creationDate: Date;
+  /**
+   * Intended to be the users symmetric key that is encrypted in some form, the current way to encrypt this is with
```
Comment (JaredSnider-Bitwarden) [inline libs/common/src/abstractions/devices/responses/protected-device.response.ts:24]: Nit: `encrypt` --> `decrypt`

File: libs/common/src/auth/services/device-trust-crypto.service.implementation.ts [hunk 4]
```diff
@@ -82,6 +87,60 @@ export class DeviceTrustCryptoService implements DeviceTrustCryptoServiceAbstrac
...
+    const encryptedDevicePublicKey = await this.encryptService.encrypt(
+      decryptedDevicePublicKey,
+      newUserKey
+    );
+
+    const currentDeviceUpdateRequest = new DeviceKeysUpdateRequest();
+    currentDeviceUpdateRequest.encryptedUserKey = encryptedNewUserKey.encryptedString;
+    currentDeviceUpdateRequest.encryptedPublicKey = encryptedDevicePublicKey.encryptedString;
+
+    // TODO: For device management, allow this method to take an array of device ids that can be looped over and individually rotated
```
Comment (JaredSnider-Bitwarden) [inline libs/common/src/auth/services/device-trust-crypto.service.implementation.ts:133]: Legend.

File: libs/common/src/abstractions/devices/responses/protected-device.response.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+export class ProtectedDeviceResponse extends BaseResponse {
+  constructor(response: Jsonify<ProtectedDeviceResponse>) {
+    super(response);
+    this.id = this.getResponseProperty("id");
+    this.name = this.getResponseProperty("name");
+    this.identifier = this.getResponseProperty("identifier");
+    this.type = this.getResponseProperty("type");
+    this.creationDate = new Date(this.getResponseProperty("creationDate"));
+    this.encryptedUserKey = this.getResponseProperty("encryptedUserKey");
+    this.encryptedPublicKey = this.getResponseProperty("encryptedPublicKey");
```
Comment (JaredSnider-Bitwarden) [inline libs/common/src/abstractions/devices/responses/protected-device.response.ts:15]: These two keys should be converted to `EncString` here, right? 

See `TrustedDeviceUserDecryptionOptionResponse` for what I did previously (based on seeing this pattern elsewhere).

File: libs/common/src/auth/services/device-trust-crypto.service.implementation.ts [hunk 6]
```diff
@@ -82,6 +87,60 @@ export class DeviceTrustCryptoService implements DeviceTrustCryptoServiceAbstrac
...
+
+    // Get the keys that are used in rotating a devices keys from the server
+    const currentDeviceKeys = await this.devicesApiService.getDeviceKeys(
+      deviceIdentifier,
+      secretVerificationRequest
+    );
+
+    // Decrypt the existing device public key with the old user key
+    const decryptedDevicePublicKey = await this.encryptService.decryptToBytes(
+      new EncString(currentDeviceKeys.encryptedPublicKey),
```
Comment (JaredSnider-Bitwarden) [inline libs/common/src/auth/services/device-trust-crypto.service.implementation.ts:113]: Note: if you convert the `ProtectedDeviceResponse` to create `EncString` then you won't need to do this here.

File: libs/common/src/auth/services/device-trust-crypto.service.implementation.ts [hunk 7]
```diff
@@ -82,6 +87,60 @@ export class DeviceTrustCryptoService implements DeviceTrustCryptoServiceAbstrac
     return deviceResponse;
   }
 
+  async rotateDevicesTrust(newUserKey: UserKey, masterPasswordHash: string): Promise<void> {
```
Comment (JaredSnider-Bitwarden) [inline libs/common/src/auth/services/device-trust-crypto.service.implementation.ts:90]: Double checked logic against our tech breakdown, and it looks perfect!

File: libs/common/src/auth/services/device-trust-crypto.service.spec.ts [hunk 8]
```diff
@@ -454,5 +457,110 @@ describe("deviceTrustCryptoService", () => {
         expect(setDeviceKeySpy).toHaveBeenCalledWith(null);
       });
     });
+    // TOOD Add tests for decryptUserKeyWithDeviceKey when types are available
```
Comment (JaredSnider-Bitwarden) [inline libs/common/src/auth/services/device-trust-crypto.service.spec.ts:460]: This comment should be removed. I think I already did once.

File: libs/common/src/auth/services/device-trust-crypto.service.spec.ts
Comment (JaredSnider-Bitwarden) [inline libs/common/src/auth/services/device-trust-crypto.service.spec.ts:1]: Awesome work testing the new function and key rotation logic!

However, you might want to review the https://contributing.bitwarden.com/contributing/testing/unit/naming-conventions/ for the test names' adherence to the latest naming conventions. I just learned about those myself recently!

File: libs/common/src/auth/services/device-trust-crypto.service.spec.ts [hunk 9]
```diff
@@ -454,5 +457,110 @@ describe("deviceTrustCryptoService", () => {
...
+              })
+            );
+          });
+
+          // Mock the decryption of the public key with the old user key
+          encryptService.decryptToBytes.mockImplementationOnce((_encValue, privateKeyValue) => {
+            expect(privateKeyValue.key.byteLength).toBe(64);
+            expect(new Uint8Array(privateKeyValue.key)[0]).toBe(5);
+            const data = new Uint8Array(new ArrayBuffer(250));
+            data.fill(17, 0, 1);
```
Comment (JaredSnider-Bitwarden) [inline libs/common/src/auth/services/device-trust-crypto.service.spec.ts:527]: Nit: you could put 17 into a named variable for easier testing clarity vs testing for magic num 17 later on

File: libs/common/src/auth/services/device-trust-crypto.service.spec.ts [hunk 10]
```diff
@@ -454,5 +457,110 @@ describe("deviceTrustCryptoService", () => {
...
+          fakeOldUserKeyData.fill(5, 0, 1);
+
+          // Mock the retrieval of a user key that differs from the new one passed into the method
+          stateService.getUserKey.mockResolvedValue(
+            new SymmetricCryptoKey(fakeOldUserKeyData) as UserKey
+          );
+
+          appIdService.getAppId.mockResolvedValue("test_device_identifier");
+
+          devicesApiService.getDeviceKeys.mockImplementation((deviceIdentifier, secretRequest) => {
```
Comment (JaredSnider-Bitwarden) [inline libs/common/src/auth/services/device-trust-crypto.service.spec.ts:501]: TIL about `mockImplementation`! TY!

**Review by JaredSnider-Bitwarden — COMMENTED** [review]
Excellent work on this! Few minor questions.

File: libs/common/src/abstractions/devices/responses/protected-device.response.ts (hunk 3 above)
Comment (justindbaur) [reply libs/common/src/abstractions/devices/responses/protected-device.response.ts:24]: Are you sure about that? This is a copy of the comment on server. This references the public key which is only used for encryption.

File: libs/common/src/auth/services/device-trust-crypto.service.spec.ts (hunk 8 above)
Comment (justindbaur) [reply libs/common/src/auth/services/device-trust-crypto.service.spec.ts:460]: Maybe something weird with the merge. Removed it.

File: libs/common/src/auth/services/device-trust-crypto.service.spec.ts
Comment (justindbaur) [reply libs/common/src/auth/services/device-trust-crypto.service.spec.ts:1]: Updated the names, I wasn't totally sure how to name my "inner" describe, I feel like it doesn't read quite as well now the the `when` condition coming after but I did my best and let me know if you have any ideas.

File: libs/common/src/abstractions/devices/responses/protected-device.response.ts (hunk 3 above)
Comment (JaredSnider-Bitwarden) [reply libs/common/src/abstractions/devices/responses/protected-device.response.ts]: Oh, I misread this. I'm sorry!

**Review by JaredSnider-Bitwarden — APPROVED** [review]
Awesome work!

**Review by djsmith85 — APPROVED** [review]
Approving the file moves I requested!

**Pull request merged**